streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -714,13 +717,26 @@ void shutdown(final boolean clean) {
             }
         }
 
-        if (clean && !consumedOffsetsAndMetadataPerTask.isEmpty()) {
-            commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
+        try {
+            if (clean && !consumedOffsetsAndMetadataPerTask.isEmpty()) {
+                commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
```

Comment:
Thinking about punctuation, should we actually call `commitOffsetsOrTransaction()` unconditionally (ie, not consider if `consumedOffsetsAndMetadataPerTask` is empty or not?

We can still move the check inside `consumedOffsetsAndMetadataPerTask`, but for EOS there might pending writes from punctuation that we still need to commit?

Reply:
Hm. So in the punctuation case -- where `commitNeeded` is true but `consumedOffsets` is empty -- we still need to call `commitOffsetsOrTransaction` (and `postCommit`) because the punctuation may for example write to a state store and generate changelog records. So we would need to commit that transaction, and also write the checkpoint file.
Makes sense

Reply:
Committable offsets here should contain consumed offsets, and punctuation itself should never update those consumed offsets right?

I think we can skip the call if `consumedOffsetsAndMetadataPerTask` is empty.

Reply:
> Committable offsets here should contain consumed offsets, and punctuation itself should never update those consumed offsets right

Yes.

> I think we can skip the call if consumedOffsetsAndMetadataPerTask is empty.

For non-eos, yes, because not non eos `commitOffsetsOrTransaction()` would only commit offsets. However, for eos (alpha and beta), we might have a pending transaction that we need to commit.